OPAL/br/src/main/scala/org/opalj/br/fpcf/properties/fieldaccess/MethodFieldAccessInformation.scala

```diff
@@ -0,0 +1,252 @@
+    )(implicit declaredFields: DeclaredFields): Iterator[DeclaredField] = {
+        val indirectAccessedFields = _indirectAccessedReceiversByField.get(accessContext.id).flatMap(_.get(pc)).getOrElse(IntMap.empty).keysIterator
+        indirectAccessedFields.take(n).map(declaredFields.apply)
+    }
+
+    def indirectAccessReceiver(
+        accessContext: Context, pc: PC, field: DeclaredField
+    ): AccessReceiver = {
+        _indirectAccessedReceiversByField(accessContext.id)(pc)(field.id)
+    }
+
+    def indirectAccessParameter(
+        accessContext: Context,
+        pc:            PC,
+        field:         DeclaredField
+    ): AccessParameter = {
+        _indirectAccessedParametersByField(accessContext.id)(pc)(field.id)
+    }
+
+    def numDirectAccesses: Int = _directAccessedFields.size
+
+    private def numIndirectAccessesInAllAccessSites: Int =
+        _indirectAccessedReceiversByField.valuesIterator.map { _.valuesIterator.map { _.size }.sum }.sum
+
+    def numIndirectAccessSites(accessContext: Context): Int = _indirectAccessedReceiversByField.get(accessContext.id).map(_.size).getOrElse(0)
+
+    def numIndirectAccesses(accessContext: Context, pc: PC): Int = _indirectAccessedReceiversByField.get(accessContext.id).flatMap(_.get(pc)).map(_.size).getOrElse(0)
+
+    def numIncompleteAccessSites: Int = _incompleteAccessSites.valuesIterator.map { _.size }.sum
+
+    def checkIsEqualOrBetterThan(e: Entity, other: Self): Unit = {
+        if (numDirectAccesses > other.numDirectAccesses ||
+            numIndirectAccessesInAllAccessSites > other.numIndirectAccessesInAllAccessSites) {
+            throw new IllegalArgumentException(s"$e: illegal refinement of $other to $this")
+        }
+    }
+
+    protected def integrateDirectAccessInformationForContext(
+        baseMap:     IntMap[IntMap[IntTrieSet]],
+        contextId:   Int,
+        updateValue: IntMap[IntTrieSet]
+    ): IntMap[IntMap[IntTrieSet]] = {
+        baseMap.updateWith(
+            contextId,
+            updateValue,
+            (o, n) => o.unionWith(
+                n,
+                (_, l, r) => {
+                    if (l == r) l
+                    else throw new UnknownError("Incompatible accessed fields for direct call")
+                }
+            )
+        )
+    }
+
+    protected def integrateIndirectAccessInformationForContext[AIT](
+        baseMap:     IntMap[IntMap[IntMap[AIT]]],
+        contextId:   Int,
+        updateValue: IntMap[IntMap[AIT]],
+        failure:     () => Throwable
+    ): IntMap[IntMap[IntMap[AIT]]] = {
+        baseMap.updateWith(
+            contextId,
+            updateValue,
+            (o, n) =>
+                o.unionWith(
+                    n,
+                    (_, l, r) => {
+                        r.unionWith(
+                            l,
+                            (_, vl, vr) =>
+                                if (vl == vr) vl
+                                else throw failure()
```

Comment:
Because this one throws different errors for different types of mismatches for indirect accesses (receiver or parameter), whereas the other one only records accessed fields for direct accesses.